plugins/output/adapt/importsource.js

```diff
@@ -0,0 +1,640 @@
+  * with matching filename is not found in the database.
+  */
+  function addAssets(assetTags, done) {
+
+    async.eachSeries(assetFolders, function iterator(assetDir, doneAssetFolder) {
+      var assetDirPath = path.join(COURSE_JSON_PATH, COURSE_LANG, assetDir);
+
+      if (fs.existsSync(assetDirPath)) {
+        var assetsGlob = path.join(COURSE_JSON_PATH, COURSE_LANG, assetDir, '*');
+        var assetsJsonFilename = path.join(COURSE_JSON_PATH, COURSE_LANG, Constants.Filenames.Assets);
+
+        var assetsJson = {};
+
+        if (fs.existsSync(assetsJsonFilename)) {
+          assetsJson = fs.readJSONSync(assetsJsonFilename);
+        }
+
+        glob(assetsGlob, function (error, assets) {
+          if(error) {
+            return cb(error);
+          }
+          var repository = configuration.getConfig('filestorage') || 'localfs';
+          async.eachSeries(assets, function iterator(assetPath, doneAsset) {
+            if (error) {
+              return doneAsset(error);
+            }
+            var assetName = path.basename(assetPath);
+            var assetExt = path.extname(assetPath);
+            var assetId = path.basename(assetPath);
+            var fileStat = fs.statSync(assetPath);
+            var assetTitle = assetName;
+            var assetDescription = assetName;
+            var tags = assetTags.slice();
+
+            if (assetsJson[assetName]) {
+              assetTitle = assetsJson[assetName].title;
+              assetDescription = assetsJson[assetName].description;
+
+              assetsJson[assetName].tags.forEach(function(tag) {
+                tags.push(tag._id);
+              });
+            }
+
+            var fileMeta = {
+              oldId: assetId,
+              title: assetTitle,
+              type: mime.lookup(assetName),
+              size: fileStat["size"],
+              filename: assetName,
+              description: assetDescription,
+              path: assetPath,
+              tags: tags,
+              repository: repository,
+              createdBy: app.usermanager.getCurrentUser()._id
+            };
+
+            if(!fileMeta) {
+              return doneAsset(new helpers.ImportError('No metadata found for asset: ' + assetName));
+            }
+            helpers.importAsset(fileMeta, metadata, doneAsset);
+          }, doneAssetFolder);
+        });
+      } else {
+        logger.log('error', 'Framework import error. Cannot find folder: ' + assetDirPath);
+        doneAssetFolder();
+      }
+    }, done);
+  }
+
+  /**
+  * Installs any custom plugins
+  */
+  function installPlugins(done) {
+    async.series([
+      function mapPluginIncludes(cb) {
+        async.each(plugindata.pluginTypes, function iterator(pluginType, doneMapIterator) {
+
+          var srcDir = path.join(COURSE_ROOT_FOLDER, Constants.Folders.Source, pluginType.folder);
+          if (fs.existsSync(srcDir)) {
```

Comment:
Probably another good candidate for returning early? I get a bit lost in the 7/8 levels of indentation.
```js
if (!fs.existsSync(srcDir)) {
  logger.log('info', 'No plugins found.');
  return doneMapIterator();
}

// fs.readdDir...